internal/dialectadapter/locker.go

```diff
@@ -0,0 +1,139 @@
+package dialectadapter
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/pressly/goose/v4/internal/dialectadapter/dialectquery"
+	"github.com/sethvargo/go-retry"
+)
+
+var (
+	// defaultLockID is the id used to lock the database for migrations. It is a crc64 hash of the
+	// string "goose". This is used to ensure that the lock is unique to goose.
+	//
+	// crc64.Checksum([]byte("goose"), crc64.MakeTable(crc64.ECMA))
+	defaultLockID int64 = 5887940537704921958
+
+	// ErrLockNotImplemented is returned when the database does not support locking.
+	ErrLockNotImplemented = errors.New("lock not implemented")
+)
+
+// Locker defines the methods to lock and unlock the database.
+//
+// Locking is an experimental feature and the underlying implementation may change in the future.
+//
+// The only database that currently supports locking is Postgres. Other databases will return
+// ErrLockNotImplemented.
+type Locker interface {
+	// CanLock returns true if the database supports locking.
+	CanLock() bool
+
+	// LockSession and UnlockSession are used to lock the database for the duration of a session.
+	//
+	// The session is defined as the duration of a single connection and both methods must be called
+	// on the same connection.
+	LockSession(ctx context.Context, conn *sql.Conn) error
+	UnlockSession(ctx context.Context, conn *sql.Conn) error
+
+	// LockTransaction is used to lock the database for the duration of a transaction.
+	// LockTransaction(ctx context.Context, tx *sql.Tx) error
+}
+
+func (s *store) CanLock() bool {
+	switch s.querier.(type) {
+	case *dialectquery.Postgres:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *store) LockSession(ctx context.Context, conn *sql.Conn) error {
+	var fn func(context.Context) error
+
+	switch t := s.querier.(type) {
+	case *dialectquery.Postgres:
+		fn = func(ctx context.Context) error {
+			row := conn.QueryRowContext(ctx, t.TryAdvisoryLockSession(defaultLockID))
+			var locked bool
+			if err := row.Scan(&locked); err != nil {
+				return err
+			}
+			if locked {
+				// A session-level advisory lock was acquired.
+				return nil
+			}
+			// A session-level advisory lock could not be acquired. This is likely because another
+			// process has already acquired the lock. We will continue retrying until the lock is
+			// acquired or the maximum number of retries is reached.
+			return retry.RetryableError(errors.New("failed to acquire lock"))
+		}
+	default:
+		return ErrLockNotImplemented
+	}
+	return retry.Do(ctx, s.retryLock, fn)
```

Comment:
It would be nice to have the ability to decide if we want to retry (in other words - if migration can't just return an error when migrations are locked) or how many times we want to retry. 

For now, we can just cancel context, but we don't know if we are stuck on mock or in the mgiration.